[connector/spanmetrics] Fix spanmetrics for child spans

shivanthzen opened the pull request:
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
#### Description
Currently, metric start timestamps are associated with the parent span (resource level). This means that all child spans, even those that occur asynchronously or infrequently, inherit the same start timestamp. This can lead to inaccurate data.

This merge request proposes moving the start timestamp (and last seen timestamp) from the parent span level to the individual child span (metric) level. This will ensure that each metric has its own accurate start and last seen timestamps, regardless of its relationship to other spans.


<!-- Issue number (e.g. #1234) or full URL to issue, if applicable. -->
#### Link to tracking issue
Fixes https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/35994

<!--Describe what testing was performed and which tests were added.-->
#### Testing

<!--Describe the documentation added.-->
#### Documentation

<!--Please delete paragraphs that you did not use before submitting.-->

shivanthzen committed:
draft spanmetrics fix

shivanthzen committed:
Support deltatemporality

shivanthzen commented:
not stale

portertech commented:
@shivanthzen are you able to look at the failing tests? For example:

```
FAIL: TestTimestampsForUninterruptedStream/AGGREGATION_TEMPORALITY_CUMULATIVE
```

renovate[bot] committed:
Update module google.golang.org/grpc to v1.69.4 - abandoned (#37196)

This PR contains the following updates:

| Package | Change | Age | Adoption | Passing | Confidence |
|---|---|---|---|---|---|
| [google.golang.org/grpc](https://redirect.github.com/grpc/grpc-go) |
`v1.69.2` -> `v1.69.4` |
[![age](https://developer.mend.io/api/mc/badges/age/go/google.golang.org%2fgrpc/v1.69.4?slim=true)](https://docs.renovatebot.com/merge-confidence/)
|
[![adoption](https://developer.mend.io/api/mc/badges/adoption/go/google.golang.org%2fgrpc/v1.69.4?slim=true)](https://docs.renovatebot.com/merge-confidence/)
|
[![passing](https://developer.mend.io/api/mc/badges/compatibility/go/google.golang.org%2fgrpc/v1.69.2/v1.69.4?slim=true)](https://docs.renovatebot.com/merge-confidence/)
|
[![confidence](https://developer.mend.io/api/mc/badges/confidence/go/google.golang.org%2fgrpc/v1.69.2/v1.69.4?slim=true)](https://docs.renovatebot.com/merge-confidence/)
|

---

> [!WARNING]
> Some dependencies could not be looked up. Check the Dependency
Dashboard for more information.

---

### Release Notes

<details>
<summary>grpc/grpc-go (google.golang.org/grpc)</summary>

###
[`v1.69.4`](https://redirect.github.com/grpc/grpc-go/releases/tag/v1.69.4):
Release 1.69.4

[Compare
Source](https://redirect.github.com/grpc/grpc-go/compare/v1.69.2...v1.69.4)

### Bug Fixes

- rbac: fix support for :path header matchers, which would previously
never successfully match
([#&#8203;7965](https://redirect.github.com/grpc/grpc-go/issues/7965)).

### Documentation

- examples/features/csm_observability: update example client and server
to use the helloworld service instead of echo service
([#&#8203;7945](https://redirect.github.com/grpc/grpc-go/issues/7945)).

</details>

---

### Configuration

📅 **Schedule**: Branch creation - "on tuesday" (UTC), Automerge - At any
time (no schedule defined).

🚦 **Automerge**: Disabled by config. Please merge this manually once you
are satisfied.

♻ **Rebasing**: Whenever PR becomes conflicted, or you tick the
rebase/retry checkbox.

🔕 **Ignore**: Close this PR and you won't be reminded about this update
again.

---

- [ ] <!-- rebase-check -->If you want to rebase/retry this PR, check
this box

---

This PR was generated by [Mend Renovate](https://mend.io/renovate/).
View the [repository job
log](https://developer.mend.io/github/open-telemetry/opentelemetry-collector-contrib).

<!--renovate-debug:eyJjcmVhdGVkSW5WZXIiOiIzOS45Mi4wIiwidXBkYXRlZEluVmVyIjoiMzkuMTA3LjAiLCJ0YXJnZXRCcmFuY2giOiJtYWluIiwibGFiZWxzIjpbImRlcGVuZGVuY2llcyIsInJlbm92YXRlYm90Il19-->

---------

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>
Co-authored-by: opentelemetrybot <107717825+opentelemetrybot@users.noreply.github.com>
Co-authored-by: Yang Song <songy23@users.noreply.github.com>

shivanthzen commented:
not  stale

shivanthzen committed:
Fix startTimestamp for delta metrics

Apparently, for delta temporality, all the metrics collected are reset
to empty state after consuming one round of data. Hence storing the
lastseentimestamp within the metric doesn't work. This PR moves
lastseentimestamp  out of metric into a map (as it was stored
previously)
of

shivanthzen commented:
@portertech Sorry about the delay. Please take another look.

shivanthzen committed:
fix lint errors

shivanthzen committed:
remove unused variable

atoulme commented:
@portertech please review

atoulme committed:
Merge branch 'main' into fix_spanmetrics

atoulme commented:
Please add a changelog.

shivanthzen committed:
Add changelog

shivanthzen committed:
Merge branch 'main' into fix_spanmetrics

shivanthzen commented:
@atoulme Done

atoulme commented:
@portertech and @Frapschen please review.

Frapschen commented on .chloggen/fix_spanmetrics.yaml:
```diff
@@ -0,0 +1,22 @@
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: connector/spanmetrics
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: This change proposes moving the start timestamp (and last seen timestamp) from the parent span level to the individual child span (metric) level. This will ensure that each metric has its own accurate start and last seen timestamps, regardless of its relationship to other spans.
```
```suggestion
note: This change proposes moving the start timestamp (and last seen timestamp) from the resourceMetrics level to the individual metrics level. This will ensure that each metric has its own accurate start and last seen timestamps, regardless of its relationship to other spans.
```

shivanthzen committed:
Update .chloggen/fix_spanmetrics.yaml

Co-authored-by: Murphy Chen <minquan.chen@daocloud.io>

shivanthzen committed:
Merge branch 'main' into fix_spanmetrics

shivanthzen committed:
Merge branch 'main' into fix_spanmetrics
